Update rootless docs (#45)
theos/theos.dev: docs/Rootless.md

Practically all jailbreaks prior to iOS 15 have been 'rootful' as they work on and install files to the (root) system directories (e.g., `/usr`, `/Library`, `/Applications`, etc). Moving forward with iOS 15 and beyond, most if not all jailbreaks will have to be 'rootless' as Apple now prevents writing to the system directories with a protection called [Signed System Volume (SSV)](https://support.apple.com/guide/security/signed-system-volume-security-secd698747c9/web). There are workarounds for this, namely bindFS, but these are less than ideal.

> Note that 'rootless' does not imply lack of `root` user permissions, contrary to what its name might suggest at a first glance.

Reviewer: ```suggestion
Note that 'rootless' does not imply lack of `root` user permissions, contrary to what its name might suggest.
```

```

- The iOS 14 arm64e ABI mentioned in [arm64e Deployment](arm64e-Deployment.html) is now *required* for the relevant devices

>     - Currently, it's only possible to build for the new ABI on macOS (as only the Xcode tools can build for it)

Reviewer: ```suggestion
    - Currently, it's only possible to build for the new ABI on macOS as the necessary `ld64` changes have not been made open source
```

Author: I would still mention that it's only with Xcode just to be 100% clear

Reviewer: Okay. Maybe "Currently, it's only possible to build for the new ABI on macOS as the necessary `ld64` changes, included in Xcode, have not been made open source" ?

>         - Unfortunately, new toolchain does not support building for the old ABI. If you want to maintain support for earlier versions, you can grab the toolchain from an earlier Xcode release as specified in [arm64e Deployment](arm64e-Deployment.html) and switch between it and the newer toolchain as desired by setting the `PREFIX` variable to the older toolchain's bin (i.e., `<xcode-ver>.xctoolchain/usr/bin/`) in your project's makefile

Reviewer: ```suggestion
        - Unfortunately, this new ABI lacks backwards compatibility with the old ABI. If you want to maintain support for earlier versions, you can grab the toolchain from an earlier Xcode release as specified in [arm64e Deployment](arm64e-Deployment.html) and switch between it and the newer toolchain as desired by setting the `PREFIX` variable to the older toolchain's bin (i.e., `<xcode-ver>.xctoolchain/usr/bin/`) in your project's makefile
```

Author: Here I think is wrong, because the point isn't that the ABIs aren't inter-compatible (as that's the whole primary point), but that the new compiler cannot build for the old ABI [apparently], which is something one might expect such as being able to cross-compile for different architectures with some switches...

Reviewer: While the issue may spawn from the abilities (or lack thereof) of the newer ld64, my thought process was that Apple doesn't seem like they would allow two arm64e slices, so the ABIs would be mutually exclusive. In that case, the issue arises due to their lack of compatiblity with each other.

Author: Yeah but the solution proposed here is to build just for the older ABI for rootful and newer for rootless, rather than have two slices at once (but the newer toolchain is now incapable of producing older ABI for such thing)

Reviewer: I see your point. In that case...
```suggestion
        - Unfortunately, this newer toolchain does not support building for the old ABI. If you want to maintain support for earlier versions, you can grab the toolchain from an earlier Xcode release as specified in [arm64e Deployment](arm64e-Deployment.html) and switch between it and the newer toolchain as desired by setting the `PREFIX` variable to the older toolchain's bin (i.e., `<xcode-ver>.xctoolchain/usr/bin/`) in your project's makefile
```

    - That being said, developers on other platforms can circumvent this *if necessary*:

>         - By [using GitHub Actions to compile your tweaks](https://github.com/p0358/SilentScreenshots/blob/master/.github/workflows/build.yml) and upload them as downloadable artifacts (free for both public and private repos)

Reviewer: ```suggestion
        - By [using GitHub Actions](https://github.com/p0358/SilentScreenshots/blob/master/.github/workflows/build.yml) to compile their tweaks (free for both public and private repos)
```

        - By using a macOS virtual machine

>             - See KVM on Linux and VMware on Windows

Reviewer: Is there a reason we don't mention QEMU here?

Author: Well KVM is the hypervisor and qemu-kvm is just what is using the KVM system to run accelerated VMs, I think there's little point mentioning just QEMU alone, which without KVM wouldn't really be able to run macOS due to overhead of emulation.

But honestly the practical reason is that by searching for macOS and KVM it's the easiest to find actual working scripts to get it running. But to be fair the main project I've linked before says "Run macOS on QEMU/KVM", so it'd still be findable, so we word it like this if you really want

>             - On Linux you can use [OSX-KVM](https://github.com/kholia/OSX-KVM) script set

Reviewer: Given the questions surrounding legality of OSX VMs, I'm hesitant to provide explicit projects that provide this.

>             - If you want to reduce resources used by the VM (after installing Xcode and Theos):

Reviewer: ```suggestion
            - If you'd like to reduce resources used by the VM after installing Xcode and Theos:
```

Author: Hmm, doesn't this make it sound more like installing Xcode and Theos increases resource usage without the parentheses? The reason I added that fragment is just so that someone doesn't disable the window server before installing Xcode and stuff

Reviewer: Good point. What about
```suggestion
            - After installing the desired tools, you can optimize the VM's system resource usage by:
```

>             - Without working App Store, download full Xcode xip archive and extract it, afterwards move `Xcode.app` to `Applications` folder (you can do this graphically through Finder)

Reviewer: Given that this is part of the OSX Theos install process, I don't think it's necessary to add here.

Author: The install docs for macOS just mention that you ought to have Xcode installed, they don't mention how and instead give you the link to App Store. And that isn't going to work on the VM (unless you do extra work to get Apple ID set up, which is risky and not really worth it here when there's alternative).

So I thought mentioning this, as a difference, is worth it. The alternative would be to update the macOS install docs to mention you can install Xcode from either app store or with xip archive.

Reviewer: I think the ways one can install Xcode is out of scope for the Theos docs...all we care about is that it's required. It may be worth swapping the Xcode hyperlink in the macOS doc to the Apple developer site link, however, as 1) it's faster than the App Store and 2) follows with what you mention about decreased risk.

Author: I swapped the link to https://developer.apple.com/download/applications/ as it has links to both then:
![image](https://github.com/theos/theos.dev/assets/5182588/9456c1ce-81ec-44ef-a8d7-f58654ff9a4f)

>             - If you want to cut on resources used by the VM:

Reviewer: ```suggestion
            - If you want to reduce resources used by the VM:
```

                - [Enable SSH access in System Preferences](https://osxdaily.com/2022/07/08/turn-on-ssh-mac/)

>                 - Disable the WindowServer daemon with `sudo launchctl disable system/com.apple.WindowServer` and reboot to drop the idle CPU usage and RAM down to ~900 MB (this disables the entire graphical interface of macOS)

Reviewer: ```suggestion
                - Disable the WindowServer daemon with `sudo launchctl disable system/com.apple.WindowServer` and reboot. This will disable macOS's graphical user interface, reducing the idle CPU and RAM usage to ~900 MB
```

>                 - Reboot and use Theos over SSH

Reviewer: ```suggestion
                - Reboot and SSH in to the VM to use Theos
```

>         - By using [allemande](https://github.com/p0358/allemande) (static binary converter to old ABI)

Reviewer: Did we still want to recommend this given its issues with blocks?

Note: I can't recall if blocks were the issue or not, but I do recall their being an inherent limitation to the allemand(e) method

Author: Well it's certainly better than the abi.py method, and I added it up before the PR to remove abi.py was pushed. In practice I think it generally worked for any tweaks I've tried, unless they contained any Swift code (including Cephei). This is something I mention on the project's README there, didn't wanna move it to the docs here in case that limitation gets out-of-date or reworded.

I still think it's better than nothing, even if just for testing. And I think we'd still rather people use this over requiring `oldabi` if it works

Author: nvm I realized I mentioned the Swift thing below xD the rest of the point stands still though xd

Reviewer: Fair enough. Thanks!

>             - This is currently the best solution if you cannot use the Xcode toolchain from macOS

Reviewer: ```suggestion
            - This is currently the best solution if you cannot macOS and Xcode 
```

Author: I'll add it to the batch, but the reason I worded it this way was because the alternative is technically still using the Xcode toolchain with Darling, just that I couldn't get it to work and gave up for now...

>             - It does not work with tweaks including Swift code (including the Cephei 2.0 library)

Reviewer: ```suggestion
            - It does not work with tweaks containing Swift code (including the Cephei v2.0 library)
```

>         - By adding `oldabi` as a dependency to their package (preferably only for testing or as a last resort as it applies system-wide and may cause stability issues for users)

Reviewer: ```suggestion
        - By adding `oldabi` as a dependency to their package (preferably only for testing or as a last resort as it applies system-wide and may cause instability for users)
```

>         - By adding `oldabi` as a dependency to their package (preferably only as a last resort or for testing, as this forces download of a system-wide tool that converts the libraries on-the-fly and may cause stability issues for users)

Reviewer: ```suggestion
        - By adding `oldabi` as a dependency to their package (preferably only for testing or as a last resort as it applies system-wide and may cause stability issues for users)
```

>             - If releasing packages without `oldabi` dependency, make sure to uninstall `oldabi` from your device during testing, to avoid accidentally releasing a tweak that silently relies on it without your knowledge!

Reviewer: ```suggestion
            - If you intend to release packages without the `oldabi` dependency, make sure to uninstall `oldabi` from your device during testing to avoid accidentally releasing a tweak that silently relies on it without your knowledge!
```

>     - All of this applies only to arm64e binaries (all system binaries), does not apply to App Store apps (as they only use arm64 architecture currently). You can continue compiling tweaks for regular apps with any toolchain.

Reviewer: ```suggestion
    - **Please note**: this only applies to arm64e binaries (i.e., system binaries and libraries). It does not apply to App Store/jailbreak apps or regular CLI binaries as arm64e is disallowed due to its unstable ABI. That is, you can continue compiling tweaks for non-system apps with any toolchain and do not need to target arm64e for your apps or CLI binaries.
```

Author: I'm not sure about mentioning "jailbreak apps", as technically devs are free to compile them for arm64e ***I think***, and probably many have done so in the past if so, meaning it could lead to some wrong assumption that it's not a possibility (and then possible frustration in debugging for the dev...)

Reviewer: Hmm, yeah, I agree. Guess "disallowed" should be swapped out in favor of something less restrictive, assuming we want to mention that jailbreak apps don't require an arm64e slice.

```

- The iOS 14 arm64e ABI mentioned in [arm64e Deployment](arm64e-Deployment.html) is now *required* for the relevant devices
    - Currently, it's only possible to build for the new ABI on macOS as the necessary `ld64` changes have not been made open source
        - Unfortunately, new toolchain does not support building for the old ABI. If you want to maintain support for earlier versions, you can grab the toolchain from an earlier Xcode release as specified in [arm64e Deployment](arm64e-Deployment.html) and switch between it and the newer toolchain as desired by setting the `PREFIX` variable to the older toolchain's bin (i.e., `<xcode-ver>.xctoolchain/usr/bin/`) in your project's makefile
    - That being said, developers on other platforms can circumvent this *if necessary*:
        - By [using GitHub Actions](https://github.com/p0358/SilentScreenshots/blob/master/.github/workflows/build.yml) to compile their tweaks (free for both public and private repos)
        - By using a macOS virtual machine
            - On Linux you can use scripts built around the KVM subsystem to get it downloaded and ran

>             - On Windows you can use projects that modify VMware configuration to get it running (requires disabling Hyper-V subsystem)

Reviewer: This feels a bit too in the weeds for Theos docs ... maybe just "See KVM, QEMU, or VMware" ?

            - Without working App Store, download full Xcode xip archive and extract it, afterwards move `Xcode.app` to `Applications` folder (you can do this graphically through Finder)
            - If you want to reduce resources used by the VM:
                - [Enable SSH access in System Preferences](https://osxdaily.com/2022/07/08/turn-on-ssh-mac/)

>                 - Disable the WindowServer daemon with `sudo launchctl disable system/com.apple.WindowServer` and reboot. This will disable macOS's graphical user interface, reducing the idle CPU and RAM usage to ~900 MB

Reviewer: ```suggestion
                - Disable the WindowServer daemon with `sudo launchctl disable system/com.apple.WindowServer` and reboot. 
                  - This will disable macOS's graphical user interface, reducing the idle CPU and RAM usage to ~900 MB
```

Additional notes:
- You do *not* need to create separate package identifiers for rootful/rootless versions of the same package
    - Newer versions of rootless-compatible package managers (e.g., Sileo and Zebra) will present only the compatible version to users

>     - Cydia will display duplicated packages, but both lead to the rootful version

Reviewer: ```suggestion
    - Cydia will display duplicate packages, but both point to the rootful version
```

    - The behavior of other package managers varies and may or may not supply the correct package to users

- All non-DEBIAN items (e.g., maintainer scripts) are placed in `/var/jb`

>     - Ensure your tweak preferences are placed in `/var/jb/var/mobile/Library/Preferences` rather than the old location of `/var/mobile/Library/Preferences`, as the latter could lead to jailbreak detection in non-jailbroken state

Reviewer: This is no longer a problem, even for legacy preference bundle impls, with the proposed stage changes in theos/775.

Reviewer: Well the bundling stage is responsible for the runtime storage location of the prefs, so that would be the point of concern. To my knowledge, rootless jailbreaks include a cfprefsd hook to redirect to the new location, which should cover the NSUserDefaults code you provide with just the suite name and possibly the CFPreferencesCopyKeyList example? Am unsure about the first NSUserDefaults example. The NSDictionary method can be solved with a simple rootless conversion macro (rootless.h), so that shouldn't pose an issue.

Author: > Well the bundling stage is responsible for the runtime storage location of the prefs, so that would be the point of concern.

I mean that's package/installation concern, and that should be handled, but actual prefs plist file is then created/accessed on runtime *in some place*, after package is already installed, so it has nothing to do with bundling

> The NSDictionary method can be solved with a simple rootless conversion macro (rootless.h), so that shouldn't pose an issue.

Yes, given the developer knows about it and notices it. Hence the warning, because they might miss it or not even be aware of the preferred new prefs location, given the old one should still work in such case.

It's not one of the things where you access packages files and would immediately notice the path is broken (like stuff people'd put in `/Library/Application Support`), because here you create and access a **new** file, meaning it's very easy to overlook

Reviewer 2: `initWithSuiteName` is definitely redirected by the jailbreak (it hooks into `cfprefd`, the preferences daemon to do so) but I'm not sure about the other more elaborate patterns.

Reviewer: ```suggestion
    - Please ensure that any additional resources you provide (e.g., newly created or laid out files and directories) are stored in a prefixed path and not a rootful path, as the latter could lead to jailbreak detection in a non-jailbroken state
```

Author: Adding it to batch, but I think it should still explicitly mention `/var/mobile`. Someone could think that rootless makes tweaks unable to write in any old non-prefixed paths whatsoever previously, meanwhile /var/mobile is the only writable path that also persists into non-jb state.

And even then I'm not sure if removing explicit note of preferences is good, as it'd be the most common case of dangling files there, which someone might not immediately think of when reading the changed version...

I'd guess you don't want to necessarily mention `/var/jb` as rootless v2 or roothide are meant to not have this symlink...

Reviewer: I feel like "as the latter could lead to jailbreak detection in a non-jailbroken state" covers that case as it implies that non-prefixed paths are writable but not worth writing to given the detection implications. 

Re prefs, I believe the configuration files you may be thinking of are handled by cfprefsd or preferenceloader and thus are automatically created in the desired path. I don't believe Theos provides any logic for this save for the base Root.plist, which will be laid out in the correct directory in staging. Please correct me if I am wrong though ... haven't taken a look at the preference situation in a little while.

- This new rootless scheme only supports iOS 15+, which itself only supports newer devices
    - This means that you do not need to compile for legacy architectures (e.g., `armv7(s)` or older) if you were previously and can bump your deployment target to 15.0 when building for rootless

---

> If needed, you can conditionally check for currently used package scheme in the Makefile, in order to differentiate the build settings used:

Reviewer: ```suggestion
If desired, you can conditionally check for the package scheme in your project's Makefile to selectively apply rootless build settings:
```

Additional notes:
- You do *not* need to create separate package identifiers for rootful/rootless versions of the same package
    - Newer versions of rootless-compatible package managers (e.g., Sileo and Zebra) will present only the compatible version to users
    - Cydia will display duplicate packages, but both point to the rootful version

>         - Users can mitigate this with [cyarchfix](https://www.ios-repo-updates.com/repository/poomsmart/package/com.ps.cyarchfix/) tweak by [PoomSmart](https://github.com/PoomSmart/cyarchfix)

Reviewer: ```suggestion
        - This can be corrected with [cyarchfix](https://github.com/PoomSmart/cyarchfix) by PoomSmart
```
